### PR TITLE
Issue 4373 - BUG - Mapping Tree nodes can be created that are invalid

### PR DESCRIPTION
**Pull request by Firstyear**
Bug Description: The mapping tree is built and arranged based on
    the content of the nsslapd-parent-suffix attribute. However, it is
    possible that this value is invalid pointing at a non-existant
    suffix, or that it could be pointing at a suffix that is invalid
    in the suffix hierarchy that mapping trees expect.

    https://www.port389.org/docs/389ds/design/mapping_tree_assembly.html

    Fix Description: Rather than build the mapping tree by arranging
    nodes through the nsslapd-parent-suffix value, we should sort and build
    them through the known and defined suffix values in cn (which we already)
    rely upon to be correct. This allows stable ordering and avoids potential
    user and developer errors.

    fixes: #4373

    Author: William Brown <william@blackhats.net.au>

    Review by: ???

### CODE REVIEW

**Comment by Firstyear** [comment]
This is not ready for merge yet, the test cases are provided to start a discussion about how we should resolve the mapping tree construction issues that exist.

**Comment by Firstyear** [comment]
Relates #4373

**Comment by Firstyear** [comment]
@mreynolds389 I don't think this is ready for merge yet (there is a single failing case I need to understand but I think it's an environment issue as it fails on git master too). But since this relates closely to your current work we probably want your MT changes to be merged first and I'll rebase onto that :)

**Comment by Firstyear** [comment]
Okay, this is good to go, please check @mreynolds389 and @tbordaz, (I will wait until #4378 is merged to rebase and retest of course)

**Comment by Firstyear** [comment]
@mreynolds389 This update adds extended testing as discussed on the mailing list, including deeper nesting, broad and many suffixes, different lengths, and some extra ideas I thought of too. They all pass as well :)

**Comment by Firstyear** [comment]
@mreynolds389, @progier389, @tbordaz, @droideck review reminder - we have a customer who has been running this patch for a month with no issues. Would be good to have in 1.4.3 / 1.4.4 so they can stop running hotfix builds.

File: ldap/servers/slapd/mapping_tree.c
Comment (progier389) [inline ldap/servers/slapd/mapping_tree.c]: A very minor point: IMHO this comment is a bit cryptic and should be removed.
         It is a response to the erroneous comment you removed rather than a clear message that stands by itself.
         It will be puzzling when someone will read it later on without the removed message context.
         IMHO: You should just remove that comment:
                There is no need to justify that you lock the mapping tree when modifying its data.

Apart this minor point, LGTM

**Review by progier389 — APPROVED** [review]

File: ldap/servers/slapd/mapping_tree.c
Comment (Firstyear) [reply ldap/servers/slapd/mapping_tree.c]: That's a good point about the comment. I'll reword it to be clearer, and as much as I trust your review, I think that @mreynolds389 may want to have a look at this issue too for his ack as well 👍 

Thanks for your time!

**Comment by mreynolds389** [comment]
LGTM

File: ldap/servers/slapd/mapping_tree.c
Comment (mreynolds389) [inline ldap/servers/slapd/mapping_tree.c:921]: This might be the calloc of size 0 we see after creating an instance...

**Pull request merged**
